examples/common/src/framework.rs

```diff
@@ -290,116 +292,108 @@ fn start<E: Example>(
+        target.set_control_flow(ControlFlow::Poll);
+
+        if cfg!(feature = "metal-auto-capture") {
+            target.exit();
         };
-        match event {
-            event::Event::RedrawEventsCleared => {
-                #[cfg(not(target_arch = "wasm32"))]
-                spawner.run_until_stalled();
 
-                window.request_redraw();
-            }
+        match event {
             event::Event::WindowEvent {
-                event:
-                    WindowEvent::Resized(size)
-                    | WindowEvent::ScaleFactorChanged {
-                        new_inner_size: &mut size,
-                        ..
-                    },
+                event: WindowEvent::Resized(size),
                 ..
-            } => {
-                log::info!("Resizing to {:?}", size);
-                config.width = size.width.max(1);
-                config.height = size.height.max(1);
-                example.resize(&config, &device, &queue);
-                surface.configure(&device, &config);
+            } => resize_window(&mut example, size, &mut config, &mut surface, &device, &queue),
+            event::Event::WindowEvent {
+                event: WindowEvent::ScaleFactorChanged { scale_factor, .. },
+                ..
+            } if scale_factor > 0.0 => {
+                let new_size = PhysicalSize {
+                    width: (f64::from(config.width) * scale_factor) as u32,
+                    height: (f64::from(config.height) * scale_factor) as u32,
+                };
+                resize_window(&mut example, new_size, &mut config, &mut surface, &device, &queue);
             }
             event::Event::WindowEvent { event, .. } => match event {
                 WindowEvent::KeyboardInput {
-                    input:
-                        event::KeyboardInput {
-                            virtual_keycode: Some(event::VirtualKeyCode::Escape),
-                            state: event::ElementState::Pressed,
+                    event:
+                        KeyEvent {
+                            logical_key: Key::Named(NamedKey::Escape),
                             ..
                         },
                     ..
                 }
                 | WindowEvent::CloseRequested => {
-                    *control_flow = ControlFlow::Exit;
+                    target.exit();
                 }
                 #[cfg(not(target_arch = "wasm32"))]
                 WindowEvent::KeyboardInput {
-                    input:
-                        event::KeyboardInput {
-                            virtual_keycode: Some(event::VirtualKeyCode::R),
-                            state: event::ElementState::Pressed,
+                    event:
+                        KeyEvent {
+                            logical_key: Key::Character(s),
                             ..
                         },
                     ..
-                } => {
+                } if s == "r" => {
```

Comment:
```
event: event::KeyEvent {
    physical_key: keyboard::PhysicalKey::Code(keyboard::KeyCode::KeyR),
    state: event::ElementState::Pressed,
    ..
}
``` might be a better replacement, as previously shift was also ignored, and this would only match if caps lock and shift is off

Reply:
I'm worried here, that `physical_key` refers to the literal physical key which might be bad for people (like me) who use different keyboard layouts. For example I need to press `k` (physically) to get `r` (logically) but I will try it out.

Reply:
Ok, I could confirm it:
```rust
use winit::{
    event::{Event, KeyEvent, WindowEvent},
    event_loop::{ControlFlow, EventLoop},
    keyboard::{KeyCode, PhysicalKey, Key, NamedKey},
    platform::x11::WindowBuilderExtX11,
    window::WindowBuilder,
};

fn main() {
    let e = EventLoop::new().unwrap();
    let w = WindowBuilder::new().build(&e).unwrap();

    e.run(move |event, target| {
        target.set_control_flow(ControlFlow::Wait);

        match event {
            Event::WindowEvent { event, .. } => match event {
                WindowEvent::CloseRequested => target.exit(),
                WindowEvent::KeyboardInput {
                    event:
                        KeyEvent {
                            physical_key: PhysicalKey::Code(KeyCode::KeyR),
                            ..
                        },
                    ..
                } if c.to_lowercase() == "c" => {
                    println!("hello there");
                }
                _ => {}
            },
            _ => {}
        };
    })
    .unwrap();
}
```
using the code above, won't work if I press `r` with a different keyboard layout but using:
```rust
                WindowEvent::KeyboardInput {
                    event:
                        KeyEvent {
                            logical_key: Key::Character(c),
                            ..
                        },
                    ..
                } if c == "c" => {
                    println!("hello there");
                }
```
works! Using `shift` + `c` will not print anything, so I think this is fine.

Reply:
You can unresolve this conversation if you're still against this.